dns_cache/manager.go

```diff
@@ -0,0 +1,62 @@
+package dns_cache
+
+import (
+	"context"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/TykTechnologies/tyk/log"
+)
+
+var (
+	logger = log.Get().WithField("prefix", "dns-cache")
+)
+
+type dialContextFunc func(ctx context.Context, network, address string) (net.Conn, error)
+
+type IDnsCacheManager interface {
+	InitDNSCaching(ttl, checkInterval time.Duration)
+	WrapDialer(dialer *net.Dialer) dialContextFunc
+}
+
+type DnsCacheManager struct {
+	DnsCache *DnsCacheStorage
+}
+
+func NewDnsCacheManager() *DnsCacheManager {
+	return &DnsCacheManager{nil}
+}
+
+func (m *DnsCacheManager) WrapDialer(dialer *net.Dialer) dialContextFunc {
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		return m.doCachedDial(dialer, ctx, network, address)
+	}
+}
+
+func (m *DnsCacheManager) doCachedDial(d *net.Dialer, ctx context.Context, network, address string) (net.Conn, error) {
+	if m.DnsCache == nil {
+		return d.DialContext(ctx, network, address)
+	}
+
+	separator := strings.LastIndex(address, ":")
```

Comment:
just thinking about readability - we can split address with using `net.SplitHostPort` and use host and port vars for logic below